src/input/mod.rs

```diff
@@ -0,0 +1,490 @@
+            } else {
+                self.pyiter_is_empty = true;
+            }
+        });
+    }
+
+    fn input_time(&mut self) -> &u64 {
+        self.push_to_timely.time()
+    }
+
+    fn input_remains(&mut self) -> bool {
+        !self.pyiter_is_empty
+    }
+}
+
+pub(crate) fn pump_from_config(
+    py: Python,
+    config: Py<InputConfig>,
+    input_handle: InputHandle<u64, TdPyAny>,
+    worker_index: usize,
+    worker_count: usize,
+    resume_epoch: u64,
+) -> Box<dyn Pump> {
+    let input_config = config.as_ref(py);
+    if let Ok(kafka_config) = input_config.downcast::<PyCell<KafkaInputConfig>>() {
+        let kafka_config = kafka_config.borrow();
+        Box::new(KafkaPump::new(kafka_config, input_handle))
+    } else if let Ok(manual_config) = input_config.downcast::<PyCell<ManualInputConfig>>() {
+        let manual_config = manual_config.borrow();
+        Box::new(PyPump::new(
+            py,
+            worker_index,
+            worker_count,
+            resume_epoch,
+            manual_config,
+            input_handle,
+        ))
+    } else {
+        let pytype = input_config.get_type();
+        panic!("Unknown input_config type: {pytype}")
+    }
+}
+/// Used internally in lieu of the Python
+/// classes
+enum TimelyAction {
+    AdvanceTo(u64),
+    Emit(String),
+}
+struct KafkaConsumer {
+    rt: Runtime,
+    consumer: BaseConsumer<CustomContext>,
+    // eventually resume_epoch...
+    current_epoch: u64,
+    desired_batch_size: u64,
+    current_batch_size: u64,
+}
+
+impl KafkaConsumer {
+    pub(crate) fn new(config: PyRef<KafkaInputConfig>) -> Self {
+        let context = CustomContext;
+
+        let rt = tokio::runtime::Builder::new_current_thread()
+            .enable_all()
+            .build()
+            .unwrap();
+
+        let consumer: BaseConsumer<CustomContext> = rt.block_on(async {
+            ClientConfig::new()
+                .set("group.id", config.group_id.clone())
+                .set("bootstrap.servers", config.brokers.clone())
+                .set("enable.partition.eof", "false")
+                .set("session.timeout.ms", "6000")
+                // TODO: we don't really want false here, it's for demo purposes
+                .set("enable.auto.commit", "false")
+                //.set("statistics.interval.ms", "30000")
+                .set("auto.offset.reset", "earliest")
+                .set_log_level(RDKafkaLogLevel::Debug)
+                .create_with_context(context)
+                .expect("Consumer creation failed")
+        });
+
+        consumer
+            .subscribe(&[&config.topics])
+            .expect("Can't subscribe to specified topics");
+
+        Self {
+            rt,
+            consumer,
+            current_epoch: 0,
+            desired_batch_size: config.batch_size,
+            current_batch_size: 0,
+        }
+    }
+
+    pub fn next(&mut self) -> TimelyAction {
+        if self.current_batch_size == self.desired_batch_size {
```

Comment:
I'm not sure why but we see this for a few rounds before data starts coming in. As in, we don't really emit a `1` or `2` epoch...

Reply:
Is that due to an empty poll incrementing the epoch?

Reply:
I don't see how yet. We don't increment the batch until we get a message, and the desired_batch_size defaults to `1`, so it's unclear to me how we fall in here